### PR TITLE
Do not show WiFi configuration link if there are no WiFi devices

### PR DESCRIPTION
**Pull request by teclator**
## Problem

In #316 we added logic to hide the link for connecting to a WiFi network but only in case the WiFi was enable based on rfkill but if there are no WiFi devices at all then it does not makes sense to show the link.

## Solution

Do not show the link for connecting to a WiFi network if there are no WiFi devices.

Update: and also do not show an empty network section when there is no connection detected and no WiFi device available.

## Testing

- *Added a new unit test*
- *Tested manually*

## Screenshots

| With a network connection | Whitout network connection |
| --- | --- |
| Screenshot from 2023-01-02 11-29-04 | Screenshot from 2023-01-02 11-26-53 |
| Screenshot from 2023-01-02 11-33-33 | Screenshot from 2023-01-02 11-36-22 |

### CODE REVIEW

File: web/src/client/network/network_manager.js [hunk 1]
```diff
@@ -500,19 +502,29 @@ class NetworkManagerAdapter {
...
-      hostname: settings?.Hostname || ""
-    };
+  /*
+  * Returns whether the system is able to scan wifi networks based on rfkill and the presence of
+  * some wifi device
+  *
+  * @return {boolean}
+  */
+  wirelessScanSupported() {
+    const wEnabled = !!(this.proxies.manager?.WirelessEnabled && this.proxies.manager?.WirelessHardwareEnabled);
```
Comment (imobachgs) [inline web/src/client/network/network_manager.js:512]: It looks a little hard to read to me. I would rather prefer something like (not tested):

```javascript
if (this.proxies.manager === undefined) return false;

const { WirelessEnabled: wirelessEnabled, WirelessHardwareEnabled: wirelessHardwareEnabled} = this.proxies.manager;
const wifiDevices = Object.values(this.proxies.devices).filter(d => d.DeviceType === NM_DEVICE_TYPE_WIFI);
return WirelessEnabled && wirelessHardware && wifiDevices.length > 0;
```

So return earlier if the manager is unavailable, and make the conditions more readable. Additionally, why do you use "wireless" and "wifi"? It does not look consistent.

**Review by imobachgs — CHANGES_REQUESTED** [review]
Now that we have plenty of time, I would take the opportunity to improve a few things. For instance, if the network is not initialized, we display [display nothing](https://github.com/yast/d-installer/blob/master/web/src/components/network/Network.test.jsx#L52), and it says nothing to the user. Additionally, if you do not have any network device (e.g., the network cable is disconnected and you do not have any wireless device), you basically see an empty section, right?

File: web/src/client/network/network_manager.js (hunk 1 above)
Comment (teclator) [reply web/src/client/network/network_manager.js:512]: Originally it represented the state of rfill enablement but that is not the case anymore, so, the method and variable should be renamed to WiFi as we are not interested in other kind of devices by now.

Comment (teclator) [reply web/src/client/network/network_manager.js]: Adapted, I basically renamed the settings property and the method to hopefully a more meaningful name.

**Comment by teclator** [comment]
@imobachgs I addressed the suggestions from our previous comment, so, only deciding what to show in case that the section is completely empty.

Screenshot from 2022-12-16 11-29-26

**Comment by dgdavid** [comment]
> @imobachgs I addressed the suggestions from our previous comment, so, only deciding what to show in case that the section is completely empty.
> 

I know we (the YaST team) haven't talk about that yet, but in my opinion the summary mustn't contain an empty section. The section can be there or not, depending on the - not yet ready - configuration (depending on the product? on the architecture? both?), but no empty.

So, I'd add a _summary_ telling the user that D-Installer was not able to find neither, a wired network connection nor a network hardware for connection to a WiFi network.

**Comment by teclator** [comment]
> > @imobachgs I addressed the suggestions from our previous comment, so, only deciding what to show in case that the section is completely empty.
> 
> I know we (the YaST team) haven't talk about that yet, but in my opinion the summary mustn't contain an empty section. The section can be there or not, depending on the - not yet ready - [product|architecture|workflow] configuration, but no empty.
> 
> So, I'd add a _summary_ telling the user that D-Installer was not able to find neither, a wired network connection nor a network hardware for connection to a WiFi network.

@dgdavid @imobachgs something like this?

File: web/src/client/network/network_manager.js [hunk 2]
```diff
@@ -500,19 +503,40 @@ class NetworkManagerAdapter {
...
+  availableWifiDevices() {
+    return Object.values(this.proxies.devices).filter(d => d.DeviceType === NM_DEVICE_TYPE_WIFI);
+  }
+
+  /*
+  * Returns whether the system is able to scan wifi networks based on rfkill and the presence of
+  * some wifi device
+  *
+  * @return {boolean}
+  */
```
Comment (dgdavid) [inline web/src/client/network/network_manager.js:520]: NP: looks to me that those comments are miss-aligned.

```js
/*
*
```

should actually be

```js
/* 
 *
```

File: web/src/client/network/network_manager.js [hunk 3]
```diff
@@ -500,19 +503,40 @@ class NetworkManagerAdapter {
...
+
+    if (!manager) return false;
+    if (!(manager.WirelessEnabled && manager.WirelessHardwareEnabled)) return false;
+
+    return this.availableWifiDevices().length > 0;
   }
 
   /*
   * Returns NetworkManager general settings
+  *
```
Comment (dgdavid) [inline web/src/client/network/network_manager.js:532]: NP: Ah, I see we have a mix on comments style :) So keep them as they are :man_shrugging:

**Comment by teclator** [comment]
Screenshot from 2023-01-02 11-26-53

**Review by dgdavid — APPROVED** [review]
LGTM, thank you!

**Review by dgdavid — COMMENTED** [review]
Do not forget to update the .changes file.

File: web/src/client/network/network_manager.js (hunk 3 above)
Comment (imobachgs) [reply web/src/client/network/network_manager.js:532]: I do not think these are mixed styles, just miss-aligned comments :) And sure we have more of them. 😞

File: web/package/cockpit-d-installer.changes [hunk 4]
```diff
@@ -1,3 +1,11 @@
+-------------------------------------------------------------------
+Mon Jan  2 11:46:37 UTC 2023 - Knut Anderssen <kanderssen@suse.com>
+
+- Do not show the link to configure wifi networks when there are no
+  wifi devices and also inform the user when no network connection
+  was detected. (gh#yast/d-installer#323).
```
Comment (imobachgs) [inline web/package/cockpit-d-installer.changes:6]: np: I would remove the first dot

**Review by imobachgs — APPROVED** [review]
Just a minor comment in the changes log. Otherwise, LGTM. Thanks!

File: web/package/cockpit-d-installer.changes (hunk 4 above)
Comment (teclator) [reply web/package/cockpit-d-installer.changes]: Ups, fixed :P

**Pull request merged**